### PR TITLE
[Draft] Various test-suite speed/reliability improvements

### LINKED ISSUE #2212: TestCLI#test_control failing with "pool_capacity": 0
**Describe the bug**
> A clear and concise description of what the bug is.

`TestCLI#test_control` (`test/test_cli.rb:175`)sometimes fails with the result `"pool_capacity": 0` with following message.
Do you have any idea to pass the test such as injecting sleeping time before the assertion, and etc?

```
  5) Failure:
TestCLI#test_control [/builddir/build/BUILD/puma-4.3.3/usr/share/gems/gems/puma-4.3.3/test/test_cli.rb:175]:
Expected /{ "started_at": "\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z", "backlog": 0, "running": 0, "pool_capacity": 16, "max_threads": 16 }/ to match # encoding: ASCII-8BIT
#    valid: true
"{ \"started_at\": \"2020-03-30T18:59:23Z\", \"backlog\": 0, \"running\": 0, \"pool_capacity\": 0, \"max_threads\": 16 }".
```

**Puma config:**

> Please copy-paste your Puma config AND your command line options here.

It's not the target.

**To Reproduce**
> Please add reproduction steps here.

This issue sometimes happened for CPU architectures ppc64le or s390x on Fedora project's build system. Note this system does not provide the internet to outside in it. But the loop back network is available.

> Your issue will be solved very quickly if you can reproduce it with a "hello world" rack application. To do this, copy this into a file called `hello.ru`:

I can not reproduce it with the "hello world" rack application.

> There is also a Dockerfile available for reproducing Linux-specific issues. To use:

I have a willing to work for that, if you need it. Maybe I will try to reproduce it with [Travis multple CPU feature](https://docs.travis-ci.com/user/multi-cpu-architectures/) `arch: ppc64le` and `arch: s390x`.

**Expected behavior**
> A clear and concise description of what you expected to happen.

`TestCLI#test_control` passes.

**Desktop (please complete the following information):**
 - OS: Fedora 33 (= rawhide = development version) Linux
 - Puma Version: 4.3.3
 - Ruby Version: 2.7.0

### PR DESCRIPTION
**Pull request by wjordan**
### Description

Spike with a bunch of different changes focused on improving the speed and reliability of the test suite. The goal was just to quickly experiment with running tests faster and with fewer flaky failures, and to share some of the high-level ideas for early discussion. Next step would be to break out the most useful fixes/changes into smaller PRs to be merged separately.

#### Highlights

- Higher parallelization (increase `MT_CPU` from default 2 to 10)
- Reduce duration of various sleeps and timeouts in tests (e.g., `sleep 1` to `sleep 0.01`)
- Replace `UniquePort` implementation with port 0 in all tests (`UniquePort` has a race condition that can cause intermittent port conflicts) (#2270)

The following changes have since been broken out into separate smaller PRs:

- ~Fire `on_booted` after server starts (fixes #2212)~ (#2431)
- ~Catch more stream-interruption errors (`EPIPE`, etc) that can happen during shutdown/thread-exit races~ (#2312)
- ~Catch and handle ThreadPool::ForceShutdown errors throughout entire threadpool lifecycle~ (#2271)
- ~Add mutex for adding connections to reactor based on `queue_requests`~ (#2271 / #2377)
- ~Fix some race conditions / edge-cases in Reactor (process multiple triggers in single cycle, re-add timeout when client-request isn't ready to be added back to thread pool [issue #2282])~ (#2279)
- ~Move shell tests into Minitest class so they can be run in parallel~ (#2240, #2346, #2347)

#### Performance

* Ubuntu 2.7
  * Before: [2m 8s](https://github.com/puma/puma/runs/624252018)
  * After: [13s](https://github.com/wjordan/puma/runs/628509391)
* Windows 2.7
  * Before: [53s](https://github.com/puma/puma/runs/624252212)
  * After: [12s](https://github.com/wjordan/puma/runs/628509572)
* MacOS JRuby
  * Before: [3m 8s](https://github.com/puma/puma/runs/624252140)
  * After: [32s](https://github.com/wjordan/puma/runs/628509526)

### Your checklist for this pull request

- [ ] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [ ] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` the pull request title.
- [ ] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [ ] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [ ] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [ ] I have updated the documentation accordingly.
- [ ] All new and existing tests passed, including Rubocop.

### CODE REVIEW

File: .github/workflows/puma.yml [hunk 1]
```diff
@@ -9,6 +9,8 @@ jobs:
     env:
       CI: true
       TESTOPTS: -v
+      MT_CPU: 10
```
Comment (nateberkopec) [inline .github/workflows/puma.yml:12]: Niiice.

File: test/helper.rb [hunk 2]
```diff
@@ -44,22 +36,14 @@ def hit(uris)
...
-
 require "timeout"
 module TimeoutEveryTestCase
   # our own subclass so we never confused different timeouts
   class TestTookTooLong < Timeout::Error
   end
 
   def run(*)
-    ::Timeout.timeout(RUBY_ENGINE == 'ruby' ? 60 : 120, TestTookTooLong) { super }
+    ::Timeout.timeout(RUBY_ENGINE == 'ruby' ? 10 : 120, TestTookTooLong) { super }
```
Comment (nateberkopec) [inline test/helper.rb:46]: DELICIOUS

File: test/helper.rb [hunk 3]
```diff
@@ -145,3 +129,16 @@ def full_name
     end
   end
 end
+
+require 'openssl'
+# Silence warnings on SSLSocket initialization.
+module SSLSocketFix
```
Comment (nateberkopec) [inline test/helper.rb:135]: Can you point me to where these warnings happen currently?

File: test/helpers/integration.rb [hunk 4]
```diff
@@ -10,30 +10,30 @@ class TestIntegration < Minitest::Test
   TOKEN = "xxyyzz"
   WORKERS = 2
 
-  BASE = defined?(Bundler) ? "bundle exec #{Gem.ruby} -Ilib" :
-    "#{Gem.ruby} -Ilib"
+  BASE = "#{Gem.ruby} -rrubygems -Ilib"
```
Comment (nateberkopec) [inline test/helpers/integration.rb:13]: This might break tests for people running locally. In CI it will be fine of course. Not sure if we can do this.

File: test/shell/t1.rb [hunk 5]
```diff
@@ -1,10 +1,16 @@
-system "ruby -rrubygems -Ilib bin/puma -p 10102 -C test/shell/t1_conf.rb test/rackup/hello.ru &"
+tcp_port = 0
```
Comment (nateberkopec) [inline test/shell/t1.rb:1]: See #2240, I want to just get rid of this folder and replace with normal tests.

File: test/test_app_status.rb [hunk 6]
```diff
@@ -4,6 +4,7 @@
 
 require "puma/app/status"
 require "rack"
+require "rack/mock"
```
Comment (nateberkopec) [inline test/test_app_status.rb:7]: required but not used?

File: test/helper.rb [hunk 7]
```diff
@@ -44,22 +36,14 @@ def hit(uris)
   end
 end
 
-module UniquePort
-  def self.call
-    TCPServer.open('127.0.0.1', 0) do |server|
```
Comment (nateberkopec) [inline test/helper.rb:49]: YES

File: test/helper.rb [hunk 8]
```diff
@@ -2,15 +2,6 @@
 # Copyright (c) 2011 Evan Phoenix
 # Copyright (c) 2005 Zed A. Shaw
 
-if %w(2.2.7 2.2.8 2.2.9 2.2.10 2.3.4 2.4.1).include? RUBY_VERSION
```
Comment (nateberkopec) [inline test/helper.rb:5]: Hmmm... why though?

Comment (nateberkopec) [reply test/helper.rb:5]: Why move to lib I mean.

**Comment by nateberkopec** [comment]
Gut reaction: yes, yes, and hell yes.

Ready to discuss these individually/in 100-line chunks when you are.

File: test/helper.rb (hunk 3 above)
Comment (wjordan) [reply test/helper.rb:135]: This was a [Ruby 2.2 warning](https://github.com/puma/puma/runs/628616697)

File: test/test_app_status.rb (hunk 6 above)
Comment (wjordan) [reply test/test_app_status.rb:7]: I think this eliminated a 'circular require considered harmful' warning on [Ruby 2.2](https://github.com/puma/puma/runs/628616697).

File: test/helper.rb (hunk 8 above)
Comment (wjordan) [reply test/helper.rb:5]: I think I saw integration tests running Puma in a subprocess still encountering this issue on Ruby2.2. Plus I figured it made sense to warn anyone on relevant Ruby versions possibly affected by this issue that weren't already using the gem.
Hopefully not super important though since 2.2 support probably won't last much longer?

Comment (nateberkopec) [reply test/helper.rb:5]: Since we're doing a major release next, I'll revisit that policy. Dropping 2.2 especially would really help for SSL maintenance.

**Comment by sudheer-meka** [comment]
Hey @nateberkopec I have created a PR [2346](https://github.com/puma/puma/pull/2346), one of the check is failing, can you please check why.

**Comment by wjordan** [comment]
Returning to this draft after being idle for a while - I've updated the original description, crossing out items that have since been broken out into separate smaller PRs (and cross-referencing them). Next up I plan to rebase this PR to remove those parts that have been merged.

After that I plan to break out two final pieces into separate PRs:

- [ ] Fire `on_booted` after server starts (fixes #2212)
- [ ] Replace UniquePort implementation with port 0 in all tests (UniquePort has a race condition that can cause intermittent port conflicts) (see also #2270, which was closed in favor of the code from here)

After one more rebase, that should leave the test-suite speed improvements in this PR for review.

**Comment by nateberkopec** [comment]
Nice to have you back @wjordan :)

**Comment by nateberkopec** [comment]
@wjordan will close these for now, let me know if you want to come back to them
